addon/-private/column-tree.js

```diff
@@ -0,0 +1,783 @@
+/* global Ember */
+import EmberObject, { get, set } from '@ember/object';
+import ObjectProxy from '@ember/object/proxy';
+import { addObserver, removeObserver } from '@ember/object/observers';
+import { A as emberA } from '@ember/array';
+
+import { computed } from '@ember-decorators/object';
+import { readOnly } from '@ember-decorators/object/computed';
+
+import { scheduler, Token } from 'ember-raf-scheduler';
+
+import { getOrCreate } from './meta-cache';
+import { move, splice, mergeSort } from './utils/array';
+import { getScale, getOuterClientRect, getInnerClientRect } from './utils/element';
+import { MainIndicator, DropIndicator } from './utils/reorder-indicators';
+
+const SCROLL_THRESHOLD = 50;
+
+const DEFAULT_COLUMN_WIDTH = 100;
+const DEFAULT_MIN_WIDTH = 50;
+const DEFAULT_MAX_WIDTH = Infinity;
+
+export const RESIZE_MODE = {
+  STANDARD: 'standard',
+  FLUID: 'fluid',
+};
+
+export const FILL_MODE = {
+  EQUAL_COLUMN: 'equal-column',
+  FIRST_COLUMN: 'first-column',
+  LAST_COLUMN: 'last-column',
+};
+
+export const WIDTH_CONSTRAINT = {
+  NONE: 'none',
+  EQ_CONTAINER: 'eq-container',
+  GTE_CONTAINER: 'gte-container',
+  LTE_CONTAINER: 'lte-container',
+};
+
+class TableColumnMeta extends EmberObject {
+  @readOnly('_node.isLeaf') isLeaf;
+  @readOnly('_node.isFixed') isFixed;
+  @readOnly('_node.width') width;
+  @readOnly('_node.offsetLeft') offsetLeft;
+  @readOnly('_node.offsetRight') offsetRight;
+}
+
+class TableColumnProxy extends ObjectProxy {
```

Comment:
What is the motivation for this proxy?

Reply:
You may be able to model this as a WeakMap then, if it is a cache for each object in the system. I see that you're refactoring existing logic here, but proxies in general raise a performance and maintainability red flag for me.

Reply:
Totally, I've gone back and forth on this over time as well. Taking a step back, I'd say there are three pretty decent APIs. For context, here is the current target API: 

```hbs
{{#ember-table as |t|}}
  {{#t.body as |b|}}
    {{#b.row as |r|}}
      {{#r.cell as |value cell column row|}}

      {{!-- value === the actual value of the cell, can be a primitive. row[column.valuePath] --}}
      {{!-- cell === the cell cache object, it's a proxy as well but a different kind designed to save memory and allocations, see CellProxy --}}
      {{!-- column === the column proxy, with added meta field --}}
      {{!-- row === the row proxy, with added meta field --}}

      {{/r.cell}}
    {{/b.row}}
  {{/t.body}}
{{/ember-table}}
```

What we're talking about is how we yield the inner API to users. I think this would be made much easier with named parameters, but in the meantime we do have to yield these positionally.

We could go a more verbose route:

```hbs
{{#r.cell as |cellValue cellMeta column columnMeta row rowMeta|}}

{{/r.cell}}
```

The common case is that users are _just_ using `cellValue` and `columnValue` so this may not be too bad. The other option we had was to _reverse_ the "proxy" behavior, so that the cell, row, and column were _wrapped_ by their proxies:

```
{{#r.cell as |cell column row|}}
  {{!-- cell === the cell cache --}}
  {{!-- cell.value === the value of the cell --}}

  {{!-- column === column meta --}}
  {{!-- column.value === the column itself --}}

  {{!-- row === row meta --}}
  {{!-- row.value === the row itself --}}

{{/r.cell}}
```

Alternatively, maybe we don't need to yield all this information at the cell level? If so, where do we yield it and why? Definitely open to suggestions, fresh eyes on this one would be helpful